src/integration/integration.service.ts

```diff
@@ -0,0 +1,407 @@
+    let uniqCoupons: Coupon[] = uniqBy(coupons, 'code');
+    if (coupons.length !== uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'COUPONS: Duplicates found and deleted',
+      });
+    }
+
+    const { promotions, availablePromotions } =
+      await this.voucherifyService.getPromotions(cart, uniqCoupons);
+
+    if (!uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'No coupons applied, skipping voucherify call',
+      });
+
+      if (isClass(storeData)) {
+        storeData.setAvailablePromotions(availablePromotions);
+      }
+      return;
+    }
+
+    const deletedCoupons = uniqCoupons.filter(
+      (coupon) => coupon.status === 'DELETED',
+    );
+
+    deletedCoupons
+      .filter((coupon) => coupon.type !== 'promotion_tier')
+      .map((coupon) =>
+        this.voucherifyConnectorService.releaseValidationSession(
+          coupon.code,
+          sessionKey,
+        ),
+      );
+
+    if (deletedCoupons.length === uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'Deleting coupons only, skipping voucherify call',
+      });
+
+      if (typeof storeData?.setAvailablePromotions === 'function') {
+        storeData.setAvailablePromotions(availablePromotions);
+      }
+      return;
+    }
+
+    this.logger.debug({
+      msg: 'Attempt to apply coupons',
+      coupons: uniqCoupons,
+      id,
+      customerId,
+      anonymousId,
+    });
+
+    uniqCoupons = filterCouponsByLimit(
+      uniqCoupons,
+      getCouponsLimit(
+        this.configService.get<number>('COMMERCE_TOOLS_COUPONS_LIMIT'),
+      ),
+    );
+
+    let validatedCoupons: ValidationValidateStackableResponse =
+      await this.voucherifyConnectorService.validateStackableVouchers(
+        buildValidationsValidateStackableParamsForVoucherify(
+          uniqCoupons.filter((coupon) => coupon.status != 'DELETED'),
+          cart,
+          mapItemsToVoucherifyOrdersItems(items),
+        ),
+      );
+
+    const inapplicableCoupons = [];
+    if (validatedCoupons.valid === false) {
+      getCouponsByStatus(validatedCoupons.redeemables, 'INAPPLICABLE').forEach(
+        (coupon) => {
+          inapplicableCoupons.push(coupon);
+        },
+      );
+      const applicableCoupons = uniqCoupons
+        .filter((coupon) => coupon.status !== 'DELETED')
+        .filter(
+          (coupon) =>
+            !inapplicableCoupons
+              .map((redeemable) => redeemable.id)
+              .includes(coupon.code),
+        );
+      if (applicableCoupons.length === 0) {
+        storeData.setInapplicableCoupons(inapplicableCoupons);
+        return;
+      }
+      validatedCoupons =
+        await this.voucherifyConnectorService.validateStackableVouchers(
+          buildValidationsValidateStackableParamsForVoucherify(
+            applicableCoupons,
+            cart,
+            mapItemsToVoucherifyOrdersItems(items),
+          ),
+        );
+    }
+
+    let productsToAdd = [];
```

Comment:
If you use a pattern 
```ts
let x = [];
if(){
  x = someFnc()
}
```
then you lose information about the `x` type. You need to define it directly `let x: ProductToAdd[] = []` or use pattern:

```ts
const x = isClass(...)? someFnc() : [];
```

Reply:
` let productsToAdd: ProductToAdd[] = [];`